File: python/c22076452.py
import copy

"""
Coursework 1 - CMT120 Programming - Cardiff University
Author: Selman Tabet - https://selman.io/
"""

# Exercise 1 - Iris Species Classifier


def exercise1(SepalLen, SepalWid, PetalLen, PetalWid):
    """
    Takes dimensions of a flower and returns its species based on the provided measurements.
    @param SepalLen: Sepal Length
    @param SepalWid: Sepal Width
    @param PetalLen: Petal Length
    @param PetalWid: Petal Width
    """
    # Species decision tree
    if (PetalLen < 2.5):
        return "setosa"
    else:
        if (PetalWid < 1.8):
            if (PetalLen < 5.0):
                if (PetalWid < 1.7):
                    return "versicolor"
                else:
                    return "virginic"
            else:
                if (PetalWid >= 1.6):
                    if (SepalLen < 7.0):
                        return "versicolor"
                    else:
                        return "virginic"
                else:
                    return "virginic"
        else:
            if (PetalLen < 4.9):
                if (SepalLen < 6.0):
                    return "versicolor"
                else:
                    return "virginic"
            else:
                return "virginic"


# Exercise 2 - Dog Breeds Standards


def exercise2(breed, height, weight, male):
    """
    Returns whether the dog's measurements conforms to their respective breed's standards.
    @param breed: Dog Breed
    @param height: Dog Height
    @param weight: Dog Weight
    @param male: Whether the dog is male or not
    """
    data_table = {
        (True, "Bulldog"): (15, 50),
        (True, "Dalmatian"): (24, 70),
        (True, "Maltese"): (9, 7),
        (False, "Bulldog"): (14, 40),
        (False, "Dalmatian"): (19, 45),
        (False, "Maltese"): (7, 6)
    }
    margin = 0.1  # 10% margin, predefined.

    # Retrieve dog breed data
    height_avg, weight_avg = data_table[(male, breed)]

    # Compute boundary values
    max_height = height_avg*(1+margin)
    min_height = height_avg*(1-margin)
    max_weight = weight_avg*(1+margin)
    min_weight = weight_avg*(1-margin)

    # Check for standard compliance

    # Out-of-bound version - The one being used.
    if (height > max_height or height < min_height or weight > max_weight or weight < min_weight):
        return False
    else:
        return True

    # In-bound version - Bonus implementation (Inactive)
    if (height <= max_height and height >= min_height and weight <= max_weight and weight >= min_weight):
        return True
    else:
        return False


###### Exercise 3 Auxiliary Functions ######

def squared(x):
    """Squares a number - helper function for map()"""
    return x*x


def median(l):
    """Median element from a list"""
    sorted_l = sorted(l)  # Sort list to find median
    mid_pos = len(l)//2  # Middle position
    # Negated index to find potential second "mid point" in case of even value length by reverse indexing
    return (sorted_l[mid_pos] + sorted_l[~mid_pos])/2


def average(l):
    """Average value of a number list"""
    return sum(l)/len(l)


# Exercise 3 - Basic Statistics

def exercise3(l):
    """
    Takes a list and applies a variety of statistical operations on it.
    @param l: List of numbers
    """
    l_squared = list(map(squared, l))  # Square each element
    return [(min(l), average(l), median(l), max(l)),
            (min(l_squared), average(l_squared), median(l_squared), max(l_squared))]


# Exercise 4 - Finite-State Machine Simulator


def exercise4(trans, init_state, input_list):
    """
    Takes an FSM transition table, an initial state and an input list and returns its output list.
    @param trans: Transition Table
    @param init_state: Initial State
    @param input_list: Input List
    """
    state = init_state
    trans_table = {}
    output = []
    # Build transition table in an improved format
    for i in trans:
        trans_table[tuple(i.split("/"))] = tuple(trans[i].split("/"))
    for j in input_list:
        output.append(trans_table[(state, j)][1])  # Append output
        state = trans_table[(state, j)][0]  # Update state
    return output


# Exercise 5 - Document Stats


def exercise5(filename):
    """
    Takes a filename and returns a tuple containing the number of alpha characters, numeric characters, symbol characters, words, sentences and paragraphs.
    @param filename: Relative file path
    """
    with open(filename, "r") as f:
        lines = f.readlines()
        # Initialize lists
        alpha_chars = []
        numeric_chars = []
        symbol_chars = []
        words = []

        # Filter out empty lines
        paragraph_indices = [index for (
            index, item) in enumerate(lines) if item == '\n']

        for i in enumerate(paragraph_indices):
            # Last index case, out of bound otherwise
            if i[0] == len(paragraph_indices)-1:
                break
            if (i[1]+1 == paragraph_indices[i[0]+1]):
                paragraph_indices[i[0]] = -1
        paragraph_fixed = list(filter(lambda x: x != -1, paragraph_indices))
        paragraphs = len(paragraph_fixed) + 1
        # Append filtered strings on a per-line basis
        for i in lines:
            alpha_chars.append(''.join(filter(str.isalpha, i)))
            numeric_chars.append(''.join(filter(str.isnumeric, i)))
            symbol_chars.append(''.join(
                filter(lambda x: not str.isalnum(x) and not str.isspace(x), i)))
            words_formatted = i.strip().replace(
                "\n", "").replace("\r", "").replace("\t", "").replace("'", ".").replace("-", ".").replace(" ", ".").replace(",", ".").split(".")
            # Extending this time because we're handling splitted strings (which are lists)
            words.extend(words_formatted)

        # Filter out empty "words" for word count output
        words = list(filter(lambda x: x != '', words))

        # Combine all strings into one string for character counting.
        all_alpha = ''.join(alpha_chars)
        all_numeric = ''.join(numeric_chars)
        all_symbol = ''.join(symbol_chars)
        sentences = sum(1 for x in all_symbol if (
            x == "?") or (x == "!") or (x == "."))  # Count sentences by counting sentence-terminating symbols

    return (len(all_alpha), len(all_numeric), len(all_symbol), len(words), sentences, paragraphs)


# Exercise 6 - List Depth


def exercise6(l):
    """
    Takes a list and returns its depth.
    @param l: List
    """
    depth = 1  # As far as the first level is concerned, it's already 1
    for i in l:
        if isinstance(i, list):
            depth += 1  # At this point, there is an extra layer
            # Recursively check for sublists, and update depth depending on whether the sublist goes further in depth.
            depth = max(depth, exercise6(i))
    return depth

# Exercise 7 - Change, please


def exercise7(amount, coins):
    """
    Takes an amount and a number of coins and returns whether the amount can be made with the given number of coins.
    @param amount: Amount
    @param coins: Number of coins
    """
    # Base cases
    amt = amount*100  # Convert to cents
    if (coins == 0):  # No coins left
        if (amount == 0):  # No amount left
            return True
        else:  # Amount left but no coins left
            return False
    denominations = [200, 100, 50, 20, 10, 5, 2, 1]
    if (coins == 1):
        if (amt in denominations):  # Check if amount is a valid denomination
            return True
        else:
            return False
    # Recursive case
    for i in denominations:  # Try all denominations against the remaining amount
        # Recursively check if the remaining amount can be made with the remaining coins using the current denomination
        if (exercise7((amt-i)/100, coins-1)):
            return True
    return False


WORDLE_DIR = "test_data/wordle.txt"

global permutations  # Global variable to store permutations
permutations = []  # Initialize empty list on runtime

########## Exercise 8 helper function ##########


def generate_permutations(start, end=[]):
    """
    This function takes a string list as input and returns
    an array of all possible permutations of the string.
    @param start: The string list to be shuffled
    @param end: This is used for recursive calls.
    """
    # Default end value is an empty list, but the argument will be used for recursive calls
    if len(start) == 0:
        permutations.append("".join(end))
    else:
        for i in range(len(start)):
            # Fix the first character, and generate permutations for the rest of the string
            generate_permutations(start[:i] + start[i+1:], end + start[i:i+1])


# Exercise 8 - Five Letter Unscramble


def exercise8(s):
    """
    Takes a string and returns the number of words in the wordle.txt file that can be made from the string.
    @param s: The string to be used to construct the candidate set of words.
    """
    with open(WORDLE_DIR, "r") as f:
        words = f.read().split("\n")

    permutations.clear()  # Clear the global variable to ensure correct results
    generate_permutations(list(s))
    filtered_permutations = list(set(permutations))  # Get rid of duplicates
    permutations.clear()  # Clear the global variable for the next call

    result = []
    for i in filtered_permutations:
        for j in words:
            if j in i:  # Some permutation strings might constitute the word itself
                result.append(j)
    # Duplicates exist since some permutations might constitute the word itself, like "househ" and "hhouse"
    return len((set(result)))


def generate_wordle_set(green, yellow, gray):
    """
    Takes the green and yellow dictionaries along with a gray wordle set, and returns a set of words that match the specifications.
    @param green: Green wordle dictionary (key: position, value: letter)
    @param yellow: Yellow wordle dictionary (key: letter, value: set of letter positions)
    @param gray: Gray wordle set
    """
    with open(WORDLE_DIR, "r") as f:
        words = f.read().split("\n")
    filtered_words = set(words)

    for i in words:
        for j in i:  # Gray filter
            if j in gray:
                filtered_words.remove(i)
                break
        for k in green:  # Green filter
            if i[k] != green[k]:  # Not a match
                try:
                    filtered_words.remove(i)
                except KeyError:  # Already removed from previous filter
                    pass
                finally:
                    break  # No need to check the rest of the word
        for l in yellow:  # Yellow filter
            if l not in i:  # Yellow letter not found
                try:
                    filtered_words.remove(i)
                except KeyError:  # Already removed from previous filter
                    pass
                finally:
                    break  # No need to check the rest of the word

    result = list(filtered_words)

    # The final filter; removes all words that have yellow marked letters in what we know to be the wrong position.
    for i in filtered_words:
        for j in enumerate(i):
            try:
                # Retrieve known wrong positions of the correct letter
                letter_indices = yellow[j[1]]  # <- KeyError failure point
                # Test if the current index is in the set of known wrong positions
                if j[0] in letter_indices:  # Correct letter in the wrong position
                    result.remove(i)
                    break
            except KeyError:  # Untested letter, move on to the next one.
                continue

    return result

# Exercise 9 - Wordle Set


def exercise9(green, yellow, gray):
    """
    Takes the green and yellow dictionaries along with a gray wordle set, and returns the cardinality of the set of words that match the specifications.
    @param green: Green wordle dictionary (key: position, value: letter)
    @param yellow: Yellow wordle dictionary (key: letter, value: set of letter positions)
    @param gray: Gray wordle set
    """
    # Cardinality of the set
    return len(generate_wordle_set(green, yellow, gray))


# Exercise 10 - One Step of Wordle


def exercise10(green, yellow, gray):
    """
    Takes the green and yellow dictionaries along with a gray wordle set, and returns the set of words with the best fit for the configuration.
    @param green: Green wordle dictionary (key: position, value: letter)
    @param yellow: Yellow wordle dictionary (key: letter, value: set of letter positions)
    @param gray: Gray wordle set
    """
    words = generate_wordle_set(green, yellow, gray)  # Retrieve the wordle set
    # This is will store cardinality scores as keys to the set of words with that score
    word_scores = {}

    for v in words:
        # This will store the cardinalities of the wordle sets from all other words' configurations
        cardinalities = []
        # Copy the list to avoid modifying the original
        words_no_v = copy.deepcopy(words)
        # This allows us to scan every OTHER word in the set
        words_no_v.remove(v)

        for w in words_no_v:
            # Define original configuration using the original parameters (copied)
            word_config = {"green": copy.deepcopy(green), "yellow": copy.deepcopy(
                yellow), "gray": copy.deepcopy(gray)}

            for k in enumerate(v):  # v_1,v_2,v_3...
                if k[1] == w[k[0]]:  # v_i = w_i
                    word_config["green"][k[0]] = k[1]
                elif k[1] in w:  # In w but not in the same position
                    try:
                        word_config["yellow"][k[1]].add(k[0])
                    except KeyError:  # New letter, add to yellow dictionary
                        word_config["yellow"][k[1]] = {k[0]}
                else:  # Add to gray set
                    word_config["gray"].add(k[1])
            # Calculate cardinality of the new configuration
            word_cardinality = exercise9(
                word_config["green"], word_config["yellow"], word_config["gray"])
            # Add the cardinality to the list of cardinalities for v
            cardinalities.append(word_cardinality)

        # Calculate the score by summing cardinalities, then add it to the dictionary
        score = sum(cardinalities)
        try:  # In case a word has the same score as another
            word_scores[score].add(v)
        except KeyError:  # New score value, add to dictionary
            word_scores[score] = {v}

    lowest_score = min(word_scores.keys())  # Retrieve the lowest score key
    # Return the set of words with the lowest score
    return word_scores[lowest_score]


"""
This was my first attempt at exercise 7, which is the iterative version.
I have since implemented a recursive version, which is much more elegant.
But I'm keeping this here because I felt too bad to dump it after all the effort I put.
"""


def exercise7_iterative(amount, coins):
    """
    Takes an amount and a number of coins and returns whether the amount can be made with the given number of coins.
    @param amount: Amount
    @param coins: Number of coins
    """
    # The Lookup Table - LUT
    lookup_table = {
        (0.00, 0): 0.0,
        (0.01, 1): 0.01,
        (0.02, 1): 0.02,
        (0.05, 1): 0.05,
        (0.1, 1): 0.1,
        (0.2, 1): 0.2,
        (0.5, 1): 0.5,
        (1, 1): 1,
        (2, 1): 2
    }
    lookup_key = (amount, int(coins))
    if lookup_key in lookup_table:
        return True
    # (0,0) lookup already checked the case where they are both 0
    if coins == 0 or amount == 0:
        return False

    coin_list = [2.0, 1.0, 0.5, 0.2, 0.1, 0.05, 0.02, 0.01]
    # Index of the denomination to consider for change (coin_list[cmp])
    cmp = 0
    while (cmp < len(coin_list)):
        # Check if it's perfectly divisible by any single denomination (e.g. amt=50, coin=25)
        for coin in coin_list:
            # Convert from primary unit float to subunit (cents/pennies) integer representation
            condition_1 = int(amount*100) % int(coin*100) == 0
            # Coins must be appropriately exhausted
            condition_2 = int(amount*100) / int(coin*100) == coins
            if (condition_1 and condition_2):
                # If the amount is a multiple of the coin (and fits coin quota), it's possible to make change
                return True

        if (amount >= coin_list[cmp]):
            amount -= coin_list[cmp]
            coins -= 1
            amount = round(amount, 2)  # Python float fix (e.g 1.4-1)
            # This check would prevent an extra iteration earlier.
            if (amount, coins) in lookup_table:
                return True
            elif amount == 0:  # Revert change
                amount += coin_list[cmp]
                coins += 1
                amount = round(amount, 2)  # Python float fix (e.g 1.4-1)
                cmp += 1  # Seek lower denomination instead
                continue
            # Used the highest denomination every time and still need more coins, so, False.
            elif coins == 0:
                return False
            else:
                continue  # Keep trying with the same denomination

        else:
            cmp += 1  # Seek lower denomination
            continue
    """
    The while loop should terminate because of the following:
    - The comparator variable (cmp) is incremented in each case except the else clause.
    meaning that after going through the entire list, the loop guard will turn false.
    - In the else clause, the comparator is not incremented, but the amount and coins variables will be reduced.
     * Sufficiently low amount will trigger the outer else clause, causing a cmp increment.
     * Sufficiently low coins will trigger the inner else-if clause, exiting the function.
    """
    return True

File: js/c22076452.js
/*
Coursework 1 - CMT120 Programming - Cardiff University
Author: Selman Tabet - https://selman.io/
*/

const fs = require('fs');
const { isAnyArrayBuffer } = require('util/types');

module.exports = {

    // Exercise 1 - Iris Species Classifier
    /**
     * Takes dimensions of a flower and returns its species based on the provided measurements.
     * @param {number} SepalLen Sepal Length
     * @param {number} SepalWid Sepal Width
     * @param {number} PetalLen Petal Length
     * @param {number} PetalWid Petal Width
     * @returns {string} Species Name
     */
    exercise1: (SepalLen, SepalWid, PetalLen, PetalWid) => {
    //  Species decision tree
    if (PetalLen < 2.5){return "setosa";}
    else {
        if (PetalWid < 1.8){
            if (PetalLen < 5.0){
                if (PetalWid < 1.7){return "versicolor";}
                else {return "virginic";}
            }
            else {
                if (PetalWid >= 1.6){
                    if (SepalLen < 7.0){return "versicolor";}
                    else{return "virginic";}
                }
                else {return "virginic";}
            }
        }
        else {
            if (PetalLen < 4.9){
                if (SepalLen < 6.0){return "versicolor";}
                else {return "virginic";}
            }
            else {return "virginic";}
        }
    }
    },

    // Exercise 2 - Dog Breeds Standards
    /**
     * Returns whether the dog's measurements conforms to their respective breed's standards.
     * @param {string} breed 
     * @param {number} height 
     * @param {number} weight 
     * @param {boolean} male 
     * @returns {boolean} Conformity result
     */
    exercise2: (breed, height, weight, male) => {
        let dataTable = {
            "Male": {
                "Bulldog": {
                    "height": 15,
                    "weight": 50
                },
                "Dalmatian": {
                    "height": 24,
                    "weight": 70
                },
                "Maltese": {
                    "height": 9,
                    "weight": 7
                }
            },
            "Female": {
                "Bulldog": {
                    "height": 14,
                    "weight": 40
                },
                "Dalmatian": {
                    "height": 19,
                    "weight": 45
                },
                "Maltese": {
                    "height": 7,
                    "weight": 6
                }
            }
        }
        let margin = 0.1 // 10% margin, predefined.

        // Determine sex field
        if (male) {sexField = "Male"}
        else {sexField = "Female"}

        // Retrieve dog breed data
        let breedData = dataTable[sexField][breed]
        let weightAverage = breedData["weight"]
        let heightAverage = breedData["height"]

        // Compute boundary values
        let maxHeight = heightAverage*(1+margin)
        let minHeight = heightAverage*(1-margin)
        let maxWeight = weightAverage*(1+margin)
        let minWeight = weightAverage*(1-margin)

        // Check for standard compliance
        // Out-of-bound implementation
        if ((height > maxHeight) || (height < minHeight) || (weight > maxWeight) || (weight < minWeight)){return false;}
        else {return true;}
    },

    // Exercise 3 - Basic Statistics
    /**
     * Takes a list and applies a variety of statistical operations on it.
     * @param {object} l Array of numbers.
     * @returns {object} Array of arrays containing the results of the operations.
     */
    exercise3: (l) => {
        let lSquared = l.map(x => x**2);
        let lAverage = arrAverage(l)
        let lSquaredAverage = arrAverage(lSquared)
        // Apply max and min functions to arrays
        let lMax = l.reduce(maxArray, -Infinity);
        let lMin = l.reduce(minArray, Infinity);
        let lMaxSquared = lSquared.reduce(maxArray, -Infinity);
        let lMinSquared = lSquared.reduce(minArray, Infinity);
        let result = [[lMin, lAverage, median(l), lMax],
        [lMinSquared, lSquaredAverage, median(lSquared), lMaxSquared]];
        return result;
    },

    // Exercise 4 - Finite-State Machine Simulator
    /**
     * Takes an FSM transition table, an initial state and an input list and returns its output list.
     * @param {object} trans Transition Table
     * @param {string} init_state Initial State
     * @param {object} input_list Input Array
     * @returns {object} Output Array
     */
    exercise4: (trans, init_state, input_list) => {
        let state = init_state; // Set initial state
        let result = [];
        for(let i=0; i<input_list.length; i++){
            let input = input_list[i];
            let lookup = `${state}/${input}`; // Construct lookup key
            let output = trans[lookup]; // Retrieve output from transition table
            let outputSplit = output.split("/");
            state = outputSplit[0]; // Update state
            result.push(outputSplit[1]); // Add output to result
        }
        return result;
    },

    // Exercise 5 - Document Stats
    /**
     * Takes a filename and returns a tuple containing the number of alpha characters, numeric characters, symbol characters, words, sentences and paragraphs.
     * @param {string} filename Relative file path
     * @returns {object} An array containing the number of alpha characters, numeric characters, symbol characters, words, sentences and paragraphs.
     */
    exercise5: (filename) => {
        const data = fs.readFileSync(filename, "utf8");
        let lines = data.split('\n');
        // Initialize counters
        let letterCounter = 0;
        let digitCounter = 0;
        let symbolCounter = 0
        let wordCounter = 0;
        let sentenceCounter = 0;
        for (let i=0; i<lines.length; i++){
            // Remove all non-letter characters (negated alpha set) and measure length
            let lettersOnly = lines[i].replace(/[^a-zA-Z]+/g, '').length;
            let numbersOnly = lines[i].replace(/[^0-9]+/g, '').length;
            let symbolsOnly = lines[i].replace(/[^-'".,]+/g, '').length;
            let sentencesOnly = lines[i].replace(/[^.!?]+/g, '').length;
            var wordsOnly = 0;
            if (lines[i].length != 0){
                wordsOnly = lines[i].match(/(\w+)/g).length;
            }
            // Update counters
            letterCounter += lettersOnly;
            digitCounter += numbersOnly;
            symbolCounter += symbolsOnly;
            wordCounter += wordsOnly;
            sentenceCounter += sentencesOnly;
        }
        var paragraphsCounter = 0;
        var flag = false; // Flag to check if the last paragraph was counted
        for (let j=0; j<lines.length; j++){
            if ((lines[j].length == 0) && (flag == false)){ // New paragraph condition
                paragraphsCounter += 1;
                flag = true;
                continue;
            }
            if (lines[j].length == 0){ // Empty line but not a new paragraph
                continue;
            }
            else{
                flag = false; // Reset flag for next paragraph
            }
        }
        if (flag == false){
            // In case the last paragraph was not counted due to lack of empty line trailing the last paragraph
            paragraphsCounter += 1;
        }
        let result = [letterCounter, digitCounter, symbolCounter, wordCounter, sentenceCounter, paragraphsCounter];
        return result;
    },

    // Exercise 6 - List Depth
    /**
     * Takes a list and returns its depth.
     * @param {object} l List
     * @returns {number} Depth of the list
     */
    exercise6: (l) => {
        var depth = 1  // As far as the first level is concerned, it's already 1
        for (let i=0; i<l.length; i++){
            if (typeof l[i] == 'object'){
                depth += 1  // At this point, there is an extra layer
                // Recursively check for sublists, and update depth depending on whether the sublist goes further in depth.
                depth = Math.max(depth, module.exports.exercise6(l[i]))
            }
        }
        return depth
    },

    // Exercise 7 - Change, please
    /**
     * @param {number} amount Amount
     * @param {number} coins Number of coins
     * @returns {boolean} Whether the amount can be made with the given number of coins.
     */
    exercise7: (amount, coins) => {
        let amt = amount*100; // Convert amount to cents
        // Base cases
        if (coins==0){ // No coins left
            if (amount==0){ // No money and no coins
                return true;
            }
            else{ // Amount left but no coins left
                return false;
            }
        }
        let denominations = [200, 100, 50, 20, 10, 5, 2, 1] // Denominations list
        if (coins == 1){
            if (denominations.includes(amt)){ // Check if amount is one of the denominations
                return true;
            }
            else{
                return false;
            }
        }
        for (let i=0; i<denominations.length; i++){ // Try all denominations against the remaining amount
            // Recursively check if the remaining amount can be made with the remaining coins using the current denomination
            if (module.exports.exercise7((amt-denominations[i])/100, coins-1)){
                return true;
            }
        }
        return false;
    },

    // Exercise 8 - Five Letter Unscramble
    /**
     * @param {string} s The string to be used to construct the candidate set of words.
     * @returns {number} The number of words in the wordle.txt file that can be made from the string.
     */
    exercise8: (s) => {
        const data = fs.readFileSync("test_data/wordle.txt", "utf8");
        let words = data.split('\n');
        
        let permutationsArray = permutations(s);

        let result = new Set();
        permutationsArray.forEach(value => {
            for (let k=0; k<words.length; k++){
                if (value.includes(words[k])){ // Check if word is one of the permutations
                    result.add(words[k]);
                }
            }
        })
        return result.size;
    },

    // Exercise 9 - Wordle Set
    /**
     * Takes the green and yellow objects along with a gray wordle set, and returns the cardinality of the set of words that match the specifications.
     * @param {object} green Green object (key: position, value: letter)
     * @param {object} yellow Yellow object (key: letter, value: set of letter positions)
     * @param {Set<string>} gray Gray letter set
     * @returns {number} Cardinality of the wordle set
     */
    exercise9: (green, yellow, gray) => {
        words = generateWordleSet(green,yellow,gray);
        return words.size;
    },

    // Exercise 10 - One Step of Wordle
    exercise10: (green, yellow, gray) => {
        return undefined;
    },
}

/*
############################################ 
            Auxiliary functions
############################################
*/

// ----- Exercise 3 helper functions ----- //
function median(array) {
    let sortedArray = array.sort((a, b) => a - b)
    let arrayLength = sortedArray.length
    if (arrayLength % 2 == 0) { // Two midpoint values to even out
        return (sortedArray[arrayLength/2] + sortedArray[arrayLength/2 - 1])/2
    }
    else {
        return sortedArray[Math.floor(arrayLength/2)]
    }
}

function maxArray(total,item) {
    return Math.max(total,item);
}

function minArray(total,item) {
    return Math.min(total,item);
}

function sumArray(total, item) {
    return total + item;
  }

function arrAverage(array) {
    return array.reduce(sumArray)/array.length;
}


// ----- Exercise 8 helper function ----- //
/**
 * This function takes a string as input and returns an array of all possible permutations of the string.
 * @param {string} input The string to be shuffled
 * @returns {object} Array of all possible permutations of the string
 */
function permutations(input) {
    var results = [];
    if (input.length === 1) { // Base case
        results.push(input);
        return results;
    }
    for (let i = 0; i < input.length; i++) {
        let firstChar = input[i]; // Fix current character
        let charsRemaining = input.substring(0, i) + input.substring(i + 1); // Slice the remaining characters
        let innerPermutations = permutations(charsRemaining); // Recursively call permutations on the remaining characters
        for (let j = 0; j < innerPermutations.length; j++) {
            results.push(firstChar + innerPermutations[j]); // Combined fixed character with permutations of remaining characters
        }
    }
    return results;
};

// ----- Exercise 9 (and 10) helper function ----- //
/**
 * Takes the green and yellow dictionaries along with a gray wordle set, and returns a set of words that match the specifications.
 * @param {object} green Green object (key: position, value: letter)
 * @param {object} yellow Yellow object (key: letter, value: set of letter positions)
 * @param {Set<string>} gray Gray letter set
 * @returns {Set<string>}
 */
function generateWordleSet(green, yellow, gray) {
    const data = fs.readFileSync("test_data/wordle.txt", "utf8");
    let words = new Set(data.split('\n')); // To be used throughout the colored filtration process
    let wordsArray = Array.from(words); // Will be iterated over to compare and remove words from the "words" Set object

    let greenIndices = Object.keys(green); // To be used to check for green letters in words
    let yellowLetters = Object.keys(yellow); // To be used to check for yellow letters in words

    for (let i=0; i<wordsArray.length; i++){
        let word = wordsArray[i]; // Current word to be checked

        let grayFilter = false; // Gray filter flag
        for (let j = 0; j<word.length; j++){
            if (gray.has(word[j])){ // Wrong letter ==> Wrong word
                grayFilter = true; // Flag set for word removal
                break;
            }
        }
        if (grayFilter){ // Word removal
            words.delete(word);
            continue;
        }

        let yellowFilter = false; // Yellow filter flag
        yellowLetters.forEach(yellowLetter => { // Check all yellow letters
            if (!word.includes(yellowLetter)){ // Yellow letter not present in word ==> Wrong word
                yellowFilter = true; // Flag set for word removal
            }
        })
        if (yellowFilter){ // Word removal
            words.delete(word);
            continue;
        }

        let greenFilter = false; // Green filter flag
        for (let j=0; j<greenIndices.length; j++){ // Check all green indices
            if (word[greenIndices[j]] != green[greenIndices[j]]){ // Mismatched letters in the same index
                greenFilter = true; // Flag set for word removal
                break;
            }
        }

        if (greenFilter){ // Word removal
            words.delete(word);
            continue;
        }
    }

    // The final filter; removes all words that have yellow marked letters in what we know to be the wrong position.
    let result = words;
    words.forEach(word => {
        for (let i=0; i<word.length; i++){
            if (yellowLetters.includes(word[i])){
                let letterIndices = yellow[word[i]]; // Retrieve known wrong positions of the correct letter
                if (letterIndices.has(i)){ // Check if the current index is in the set of known wrong positions
                    result.delete(word); //Correct letter in the wrong position
                    break;
                }
            }
            else continue; // Untested letter, move on to the next one.
        }
    })
    return words;
}